src/compiler/chainrules.jl

```diff
@@ -115,28 +115,60 @@ for T_outer in (:Tuple, :NamedTuple)
+wrap_chainrules_output(dxs::AbstractArray) = map(wrap_chainrules_output, dxs)
+#=
+# As an optimisation, we can convert by `reinterpret` for bitstypes, e.g. arrays of tuples of numbers
+@inline function wrap_chainrules_output(dxs::AbstractArray{<:ChainRules.Tangent{<:Any, B}}) where {B}
+  if isbitstype(B)
+    # B is the backing type. It still contains NoTangent etc, which need converting to Nothing
+    reinterpret(wrap_chainrules_output(B), dxs)
+  else
+    map(wrap_chainrules_output, dxs)
+  end
+end
+wrap_chainrules_output(::Type{<:AbstractZero}) = Nothing
+wrap_chainrules_output(::Type{NamedTuple{L,T}}) where {L,T} = NamedTuple{L,wrap_chainrules_output(T)}
+@generated function wrap_chainrules_output(::Type{T}) where T<:Tuple
+  inner = map(wrap_chainrules_output, T.parameters)
+  :(Tuple{$(inner...)})
+end
+=#
 
 """
-    wrap_chainrules_input(x)
+    wrap_chainrules_input(dx)
 
-Convert `x` from the format Zygote uses internally to differentials types ChainRules uses.
+Convert `dx` from the format Zygote uses internally to differentials types ChainRules uses.
 """
-@inline wrap_chainrules_input(x) = x
+@inline wrap_chainrules_input(dx) = dx
 @inline wrap_chainrules_input(::Nothing) = ChainRules.ZeroTangent()
+@inline wrap_chainrules_input(::Tuple{Vararg{Nothing}}) = ChainRules.ZeroTangent()
 @inline wrap_chainrules_input(::AbstractArray{Nothing}) = ChainRules.ZeroTangent()
-@inline function wrap_chainrules_input(xs::Union{Tuple, NamedTuple})
-  xp = map(wrap_chainrules_input, xs)
-  ChainRules.Tangent{Any, typeof(xp)}(xp)
+@inline function wrap_chainrules_input(dxs::Union{Tuple, NamedTuple})
+  xp = map(wrap_chainrules_input, dxs)
+  # This produces Tangent{Any} since it does not get to see the primal, `x`.
+  ChainRulesCore.Tangent{Any, typeof(xp)}(xp)
 end
 # For mutable types, including x=Ref(1), Zygote makes Ref{Any}(::NamedTuple)
-@inline wrap_chainrules_input(x::Ref) = wrap_chainrules_input(x[])
-# Could `reinterpret` instead of broadcasting here -- TODO
-@inline wrap_chainrules_input(xs::AbstractArray{<:Ref}) = wrap_chainrules_input.(xs)
-@inline wrap_chainrules_input(xs::AbstractArray{<:Union{Nothing, <:Ref}}) = wrap_chainrules_input.(xs) # no test invented for this
-@inline wrap_chainrules_input(xs::AbstractArray{<:NamedTuple}) = wrap_chainrules_input.(xs)
-@inline wrap_chainrules_input(xs::AbstractArray{<:Union{Nothing, <:NamedTuple}}) = wrap_chainrules_input.(xs)
+@inline wrap_chainrules_input(dx::Ref) = wrap_chainrules_input(dx[])
+# For arrays, whitelist the safe ones, but always look inside Any[]:
+@inline wrap_chainrules_input(dxs::AbstractArray{<:Number}) = dxs
+@inline wrap_chainrules_input(dxs::AbstractArray{<:AbstractArray{<:Number}}) = dxs
+@inline wrap_chainrules_input(dxs::AbstractArray) = map(wrap_chainrules_input, dxs)
+# Could `reinterpret` instead here...
+# One easy case, but can this go wrong?
+# @inline wrap_chainrules_input(xs::Base.ReinterpretArray{<:NamedTuple, <:Tangent}) = parent(@show xs)
```

Comment:
This seems like debugging code

Reply:
It's a comment, about further optimisations? Again 1112 is the issue.

Reply:
We can get rid of the show either way. In general, it's best to leave reinterpreted arrays be, they can have unexpected behavior if we try to manage them manually. Even in a comment, I still doubt the show was intentionally expected to be part of committed code?